### PR TITLE
scripts: twister: Fix warnings & update some style (dependency injection/type hints).

### PR DESCRIPTION
**Pull request by ithinuel**
These change allow `ZEPHYR_BASE=$(pwd) PYTHON_PATH=./scripts/tests pytest ./scripts/tests/twister` to run locally from zephyr's directory to pass and do so without warning.

Please refer to each commit messages for a short explanation of the related change.

I cannot get `ZEPHYR_BASE=$(pwd) PYTHON_PATH=./scripts/tests pytest ./scripts/tests/twister_blackbox` to run locally but I lack the time to investigate what am I doing wrong.
But it passes in CI :)

### CODE REVIEW

File: scripts/pylib/twister/twisterlib/testinstance.py [hunk 1]
```diff
@@ -13,6 +15,7 @@
 import glob
 import csv
 
+from pylib.twister.twisterlib.environment import TwisterEnv
```
Comment (nashif) [inline scripts/pylib/twister/twisterlib/testinstance.py:18]: why the additional pylib.twister here?

**Review by nashif — DISMISSED** [review]

File: scripts/pylib/twister/twisterlib/handlers.py [hunk 2]
```diff
@@ -3,11 +3,15 @@
...
 # Copyright 2022 NXP
+# Copyright (c) 2024 Arm Limited (or its affiliates). All rights reserved.
+#
 # SPDX-License-Identifier: Apache-2.0
 
+import argparse
 import logging
 import math
 import os
+from typing import Optional
```
Comment (LukaszMrugala) [inline scripts/pylib/twister/twisterlib/handlers.py:14]: Move to the `from X import Y` section a bit below.

File: scripts/pylib/twister/twisterlib/harness.py [hunk 3]
```diff
@@ -564,7 +564,8 @@ def _update_test_status(self):
     def _parse_report_file(self, report):
         tree = ET.parse(report)
         root = tree.getroot()
-        if elem_ts := root.find('testsuite'):
+        elem_ts = root.find('testsuite')
+        if elem_ts is not None:
```
Comment (LukaszMrugala) [inline scripts/pylib/twister/twisterlib/harness.py:568]: What was wrong with the previous approach?

File: scripts/tests/twister/test_twister.py [hunk 4]
```diff
@@ -51,33 +53,39 @@ def test_incorrect_schema(filename, schema, test_data):
...
+        # Check that all DTC overlay files have been assembled in the correct order
+        assert ";".join(scenario["extra_dtc_overlay_files"]) == \
+            "overlay1;overlay2;overlay3;overlay4;overlay5;overlay6;overlay7;overlay8"
+
+        # Check that all overlay conf files have been assembled in the correct order
+        assert scenario["extra_overlay_confs"] == \
+            ["oc1.conf", "oc2.conf", "oc3.conf", "oc4.conf"]
+
+        # Check extra kconfig statements, too
+        assert scenario["extra_configs"] == ["CONFIG_FOO=y"]
```
Comment (LukaszMrugala) [inline scripts/tests/twister/test_twister.py:88]: You should be able to dedent those `assert`s once, I think.

**Review by teburd — DISMISSED** [review]

File: scripts/pylib/twister/twisterlib/harness.py (hunk 3 above)
Comment (ithinuel) [reply scripts/pylib/twister/twisterlib/harness.py:568]: Python 3.12 (at least) reports a deprecation warning about relying on "truthiness".
see 7933d2f43547531d960d002e89b26eae98fb8b62’s commit message for a quote of that message.
I am not well versed enough in python to figure where this comes from though :(

Comment (ithinuel) [reply scripts/pylib/twister/twisterlib/harness.py:568]: I guess `if (elem_ts := root.find('testsuite')) is not None:` might also work ?

File: scripts/pylib/twister/twisterlib/config_parser.py [hunk 5]
```diff
@@ -36,7 +36,7 @@ def extract_fields_from_arg_list(target_fields: set, arg_list: Union[str, list])
             # Move to other_fields
             other_fields.append(field)
 
-    return extracted_fields, " ".join(other_fields)
+    return extracted_fields, other_fields
```
[the remark this answers is not in the document]
Comment (ithinuel) [reply scripts/pylib/twister/twisterlib/config_parser.py:39]: As far as I can tell, `extra_args` is expected to be a list (in yaml) so returning it as a `list` seems ok to me.
I coulding find other places than `test_twister` that check for the content of this parameter explicitly. I may have missed some relevant location, could you point me to some of them please?

**Review by nashif — APPROVED** [review]

**Review by teburd — APPROVED** [review]

**Pull request merged**